build: add `tsdoc-doctest` ESLint rule

Planeshifter opened the pull request:
## Description

> What is the purpose of this pull request?

This pull request:

-   adds `tsdoc-doctest` ESLint rule for for linting return annotations in TSDoc example code inside of TypeScript declaration files.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves https://github.com/stdlib-js/metr-issue-tracker/issues/48.

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

Planeshifter committed:
build: add `tsdoc-doctest` ESLint rule for linting return annotations in TSDoc examples

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter committed:
build: minor refactoring

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter committed:
refactor: clean-up and handle interface case

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented on etc/eslint/plugins/typescript.js:
```diff
@@ -30,7 +30,10 @@ var plugins = [
-	'@typescript-eslint'
+	'@typescript-eslint',
+
+	// Stdlib custom rules:
```
```suggestion
	// Custom stdlib rules:
```

kgryte replied:
```suggestion
	// Custom stdlib rules:
```

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/tsdoc-doctest/test/fixtures/invalid.js:
```diff
@@ -0,0 +1,209 @@
+// MODULES //
+
+var join = require( 'path' ).join;
+var rootDir = require( '@stdlib/_tools/utils/root-dir' );
```
Is it possible to avoid this? I am specifically concerned about L52 where we assume that a file exists along a certain path within the project.

kgryte replied:
If you really need to resolve a package within the project rather than using fixtures, instead of `rootDir`, you could use `resolve` for dynamically resolving a package's main entry point. But even then, I don't think this is a good idea, as it means that this package cannot be tested independently of the greater project.

Planeshifter replied:
Refactored to use fixtures.

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/tsdoc-doctest/test/fixtures/unvalidated.js:
```diff
@@ -0,0 +1,69 @@
+// MODULES //
+
+var join = require( 'path' ).join;
+var rootDir = require( '@stdlib/_tools/utils/root-dir' );
```
Same comment.

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/tsdoc-doctest/test/fixtures/unvalidated.js:
```diff
@@ -0,0 +1,69 @@
+		'',
+		'export = noExample;'
+	].join( '\n' ),
+	'filename': join( ROOT_DIR, 'lib/node_modules/@stdlib/test/no-example/docs/types/index.d.ts' )
```
Given that this is a non-existent file, why is it necessary to use `ROOT_DIR` at all here?

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/tsdoc-doctest/test/fixtures/valid.js:
```diff
@@ -0,0 +1,184 @@
+// MODULES //
+
+var join = require( 'path' ).join;
+var rootDir = require( '@stdlib/_tools/utils/root-dir' );
```
Same comment.

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/tsdoc-doctest/test/fixtures/valid.js:
```diff
@@ -0,0 +1,184 @@
+		'*',
+		'* @example',
+		'* var v = sqrt( 4.0 );',
+		'* // returns 2',
```
Just checking: `2.0` works, as well, correct?

Planeshifter replied:
Yes. Changed the code here to use `2.0` to reflect best practice for return annotations.

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/tsdoc-doctest/test/fixtures/valid.js:
```diff
@@ -0,0 +1,184 @@
+		'',
+		'export = double;'
+	].join( '\n' ),
+	'filename': join( ROOT_DIR, 'lib/node_modules/@stdlib/math/base/special/double/docs/types/index.d.ts' )
```
Another example where `ROOT_DIR` does not seem necessary.

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/tsdoc-doctest/README.md:
```diff
@@ -0,0 +1,184 @@
+
+## Notes
+
+-   The corresponding JavaScript package must be loadable via `require('@stdlib/<package>')`
```
```suggestion
-   The corresponding JavaScript package must be loadable via `require('@stdlib/<package>')`.
```

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/tsdoc-doctest/README.md:
```diff
@@ -0,0 +1,184 @@
+-   The corresponding JavaScript package must be loadable via `require('@stdlib/<package>')`
+-   The rule skips validation if the package cannot be loaded.
+-   Examples are executed in a sandboxed VM context with limited globals.
+-   The rule validates assignment-form examples (e.g., `var x = fn(...);` followed by an annotation). It does not validate console output or expression-only forms using `// =>`.
```
Why this limitation?

Planeshifter replied:
Added support for linting annotations of form `// =>`.

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/tsdoc-doctest/README.md:
```diff
@@ -0,0 +1,184 @@
+-   The rule skips validation if the package cannot be loaded.
+-   Examples are executed in a sandboxed VM context with limited globals.
+-   The rule validates assignment-form examples (e.g., `var x = fn(...);` followed by an annotation). It does not validate console output or expression-only forms using `// =>`.
+-   The implementation path the rule uses to load the JavaScript implementation can be overridden via the `implementationPath` rule option (default: `../../lib` relative to the declaration file).
```
Not clear to me where/when this option is provided.

Planeshifter replied:
Option removed due to moving to `package.json` resolution.

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/tsdoc-doctest/lib/add_package_to_scope.js:
```diff
@@ -0,0 +1,90 @@
+var RE_DECLARE_FUNCTION = /declare\s+function\s+([a-zA-Z_$][a-zA-Z0-9_$]*)\s*[<(]/;
+var RE_DECLARE_VAR = /declare\s+var\s+([a-zA-Z_$][a-zA-Z0-9_$]*)\s*:/;
+var RE_DECLARE_CLASS = /declare\s+class\s+([a-zA-Z_$][a-zA-Z0-9_$]*)\s/;
+var RE_DECLARE_VAR_NAMESPACE = /declare\s+var\s+([a-zA-Z_$][a-zA-Z0-9_$]*)\s*:\s*[A-Z][a-zA-Z0-9_$]*/;
```
This regexp is not necessary. Just use L29. You can ignore the second capture group.

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/tsdoc-doctest/lib/add_package_to_scope.js:
```diff
@@ -0,0 +1,90 @@
+	var funcMatch;
+
+	if ( typeof pkg === 'function' ) {
+		// Try to match declare function pattern (handles generics with <):
```
Similar to L60, it would be nice to include an example in each of the comments (L50, L54, etc) which demonstrate what pattern we are attempting to match.

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/tsdoc-doctest/lib/main.js:
```diff
@@ -0,0 +1,487 @@
+var vm = require( 'vm' );
+var dirname = require( 'path' ).dirname;
+var resolve = require( 'path' ).resolve;
+var existsSync = require( 'fs' ).existsSync; // eslint-disable-line node/no-sync
```
Can't we use the stdlib equivalent?

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/tsdoc-doctest/lib/main.js:
```diff
@@ -0,0 +1,487 @@
+// MODULES //
+
+var vm = require( 'vm' );
+var dirname = require( 'path' ).dirname;
```
We can use the stdlib equivalent: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/utils/dirname

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/tsdoc-doctest/lib/main.js:
```diff
@@ -0,0 +1,487 @@
+* @returns {string} escaped string
+*/
+function escapeRegex( str ) {
+	return str.replace( RE_SPECIAL_CHARS, '\\$&' );
```
We can use the stdlib equivalent.

Planeshifter replied:
Changed.

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/tsdoc-doctest/lib/main.js:
```diff
@@ -0,0 +1,487 @@
+	var key;
+
+	escapedValue = escapeRegex( value );
+	key = [ annotationType, escapedValue ].join( '::' );
```
Is it really necessary to create a temporary array here?

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/tsdoc-doctest/lib/main.js:
```diff
@@ -0,0 +1,487 @@
+* @param {string} str - string to escape
+* @returns {string} escaped string
+*/
+function escapeRegex( str ) {
```
Don't we have a package for escaping regular expressions? Is it not sufficient?

Planeshifter replied:
Good catch! Changed to using built-in.

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/tsdoc-doctest/lib/main.js:
```diff
@@ -0,0 +1,487 @@
+*
+* @private
+* @param {Object} scope - VM scope
+* @param {any} expected - expected value to search for
```
`any` is not valid.

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/tsdoc-doctest/lib/main.js:
```diff
@@ -0,0 +1,487 @@
+* @private
+* @param {string} filepath - TypeScript declaration file path
+* @param {string} implementationPath - relative path to implementation
+* @returns {string|null} resolved implementation path or null if not found
```
Missing parens.

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/tsdoc-doctest/lib/main.js:
```diff
@@ -0,0 +1,487 @@
+	if ( lineCountCache[ str ] ) {
+		return lineCountCache[ str ];
+	}
+	matches = str.match( RE_NEWLINE );
```
This is an odd way to count lines. Why are you using `match` here and not `split( RE_EOL ).length`?

Planeshifter replied:
Addressed.

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/tsdoc-doctest/lib/main.js:
```diff
@@ -0,0 +1,487 @@
+*/
+function cleanTSDocComment( comment ) {
+	// Remove opening /** and closing */
+	var cleaned = comment.replace( /^\/\*\*/, '' ).replace( /\*\/$/, '' );
```
At L167, you use a stdlib API, but not here. Would be good to be consistent.

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/tsdoc-declarations-doctest/lib/main.js:
```diff
@@ -0,0 +1,487 @@
+
+	last = 0;
+	RE_ANNOTATION.lastIndex = 0;
+	try {
```
I am not sure why we are enclosing the entirety of the code logic in a `try` block. This is obviously not best practice. I see that the same thing was done in [`jsdoc-doctest`](https://github.com/stdlib-js/stdlib/blob/724283b0bc5d2eb8bf48491c954c6a119330c1f5/lib/node_modules/%40stdlib/_tools/eslint/rules/jsdoc-doctest/lib/main.js#L277); that should also be cleaned up. Otherwise, we are swallowing programmer errors.

kgryte replied:
@Planeshifter You marked this as resolved, but didn't make any changes.

Planeshifter replied:
As discussed, this will be addressed in a follow-up PR.

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/tsdoc-doctest/lib/main.js:
```diff
@@ -0,0 +1,487 @@
+	options = context.options[ 0 ] || {};
+
+	// Only process TypeScript declaration files:
+	if ( !filename.endsWith( '.d.ts' ) ) {
```
I wonder if we should rename this rule to make this more explicit that it won't work on TS source files. E.g., `tsdoc-declarations-doctest`, or similar.

Planeshifter replied:
Renamed.

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/tsdoc-doctest/lib/main.js:
```diff
@@ -0,0 +1,487 @@
+	options = context.options[ 0 ] || {};
+
+	// Only process TypeScript declaration files:
+	if ( !filename.endsWith( '.d.ts' ) ) {
```
Can we also be more consistent in using stdlib APIs? e.g., `string/ends-with`.

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/tsdoc-doctest/lib/main.js:
```diff
@@ -0,0 +1,487 @@
+		lineCountCache = {};
+
+		// Resolve implementation path relative to TypeScript declaration file path:
+		implementationPath = options.implementationPath || '../../lib';
```
Remind me: why don't we want to resolve the nearest `package.json` and the associated main entry point?

Planeshifter replied:
Refactored; reasoning was that for usage beyond `stdlib`, libraries might not use small-package approach but instead have several declaration files, in which case different path resolution might be preferable. But as implemented, wouldn't have been possible to accommodate lots of project layouts either, so now using the `package.json` approach.

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/tsdoc-doctest/lib/main.js:
```diff
@@ -0,0 +1,487 @@
+
+		// Try to load the implementation directly:
+		try {
+			pkg = require( implPath ); // eslint-disable-line stdlib/no-dynamic-require
```
Can we use `try-require` here?

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/tsdoc-doctest/lib/main.js:
```diff
@@ -0,0 +1,487 @@
+			}
+
+			for ( j = 0; j < examples.length; j++ ) {
+				code = examples[ j ].trim();
```
`string[/base]/trim`

kgryte reviewed (changes requested):
Left initial comments.

Planeshifter committed:
build: use built-ins, resolve via `package.json`, and avoid rootDir

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter committed:
build: use existing escapeRegExpString

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter committed:
refactor: address PR feedback

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter committed:
chore: rename lint rule

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte committed:
docs: fix comment

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/tsdoc-declarations-doctest/README.md:
```diff
@@ -0,0 +1,183 @@
+
+# tsdoc-declarations-doctest
+
+> [ESLint rule][eslint-rules] to ensure that return annotations in TSDoc examples match the actual output in TypeScript declaration files (`.d.ts`).
```
```suggestion
> [ESLint rule][eslint-rules] to ensure that return annotations in TSDoc examples match the actual output in TypeScript declaration files (`*.d.ts`).
```

kgryte committed:
docs: update description

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/tsdoc-declarations-doctest/README.md:
```diff
@@ -0,0 +1,183 @@
+
+#### rule
+
+[ESLint rule][eslint-rules] to ensure that return annotations in TSDoc examples match the actual output in TypeScript declaration files (`.d.ts`). Return annotations may start with `returns`, `throws`, or `=>`. `returns` follow variable declarations or assignment expressions, whereas `=>` follow expression-only forms including `console.log` calls. The rule validates `@example` blocks in TSDoc com ...
```
```suggestion
[ESLint rule][eslint-rules] to ensure that return annotations in TSDoc examples match the actual output in TypeScript declaration files (`*.d.ts`). Return annotations may start with `returns`, `throws`, or `=>`. `returns` follow variable declarations or assignment expressions, whereas `=>` follow expression-only forms including `console.log` calls. The rule validates `@example` blocks in TSDoc comments within `.d.ts` files by automatically resolving the corresponding implementation via the nearest `package.json` file.
```

kgryte committed:
docs: update description

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
docs: fix notes placement and update description

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/tsdoc-declarations-doctest/README.md:
```diff
@@ -0,0 +1,184 @@
+* // returns 6
+*/
+declare function add( x: number, y: number ): number;
+
+export = add;
```
```suggestion
```

kgryte committed:
docs: update example

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/tsdoc-declarations-doctest/README.md:
```diff
@@ -0,0 +1,182 @@
+*/
+declare function add( x: number, y: number ): number;
+
+export = add;
```
```suggestion
```

kgryte committed:
docs: update example

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
docs: fix comment style

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/tsdoc-declarations-doctest/lib/add_package_to_scope.js:
```diff
@@ -0,0 +1,86 @@
+
+	pkgType = typeof pkg;
+	if ( pkgType === 'function' ) {
+		match = sourceText.match( RE_DECLARE_FUNCTION ) ||
+			sourceText.match( RE_DECLARE_VAR ) ||
+			sourceText.match( RE_DECLARE_CLASS );
```
```suggestion
		match = sourceText.match( RE_DECLARE_FUNCTION ) || sourceText.match( RE_DECLARE_VAR ) || sourceText.match( RE_DECLARE_CLASS ); // eslint-disable-line max-len
```

kgryte committed:
style: don't break expressions

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/tsdoc-declarations-doctest/lib/add_package_to_scope.js:
```diff
@@ -0,0 +1,84 @@
+		if ( pkgType === 'object' && pkg !== null ) {
+			namespaceMatch = sourceText.match( RE_DECLARE_VAR_INTERFACE );
+			if ( namespaceMatch ) {
+				scope[ namespaceMatch[1] ] = pkg;
```
@Planeshifter You fall-through here. Is that intended?

kgryte replied:
If so, it would be good to document that the fall-through is intended.

Planeshifter replied:
Intentional. Comments added.

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/tsdoc-declarations-doctest/lib/main.js:
```diff
@@ -0,0 +1,498 @@
+
+var vm = require( 'vm' );
+var resolve = require( 'path' ).resolve;
+var readFileSync = require( 'fs' ).readFileSync; // eslint-disable-line node/no-sync
```
@Planeshifter We should be using `@stdlib/fs/read-file`.sync.

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/tsdoc-declarations-doctest/lib/main.js:
```diff
@@ -0,0 +1,498 @@
+	if ( lineCountCache[ str ] ) {
+		return lineCountCache[ str ];
+	}
+	lines = str.split( reEOL() );
```
@Planeshifter You don't need to create a new EOL regexp here. Use `.REGEXP` from the respective package.

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/tsdoc-declarations-doctest/lib/main.js:
```diff
@@ -0,0 +1,498 @@
+function countLines( str ) {
+	var lines;
+
+	// Use cached result if available
```
```suggestion
```

What if the line count in the cache is `0`?

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/tsdoc-declarations-doctest/lib/main.js:
```diff
@@ -0,0 +1,498 @@
+}
+
+/**
+* Searches for variable in scope matching the expected value.
```
```suggestion
* Searches for a variable in scope matching the expected value.
```

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/tsdoc-declarations-doctest/lib/main.js:
```diff
@@ -0,0 +1,498 @@
+}
+
+/**
+* Resolves the implementation path from package.json.
```
```suggestion
* Resolves the implementation path from the nearest parent `package.json`.
```

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/tsdoc-declarations-doctest/lib/main.js:
```diff
@@ -0,0 +1,498 @@
+
+	// Read and parse package.json:
+	try {
+		pkgJSON = JSON.parse( readFileSync( pkgPath, 'utf8' ) );
```
@Planeshifter Use `@stdlib/utils/parse-json`. Then handle with

```
json = parseJSON(...);
if ( json instanceof Error ) {
    debug( .... );
    return null;
}
```

But also, more generally here, you can use `@stdlib/fs/read-json` which combines both reading and parsing.

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/tsdoc-declarations-doctest/lib/main.js:
```diff
@@ -0,0 +1,498 @@
+	// Get the main entry point:
+	mainPath = pkgJSON.main || './lib';
+	pkgDir = dirname( pkgPath );
+	return resolve( pkgDir, mainPath );
```
```suggestion
	return resolve( dirname( pkgPath ), pkgJSON.main || './lib' );
```

The intermediate variables are not particularly useful here.

kgryte replied:
What if `pkgJSON.main` is the empty string? In that case, main would default to `index.js`. Here, you default to `./lib` which I don't think would be correct.

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/tsdoc-declarations-doctest/lib/main.js:
```diff
@@ -0,0 +1,498 @@
+	var cleaned = replace( replace( comment, /^\/\*\*/, '' ), /\*\/$/, '' );
+
+	// Remove * at the beginning of lines
+	cleaned = replace( cleaned, RE_COMMENT_PREFIX, '' );
+	return cleaned;
```
```suggestion
	return replace( cleaned, RE_COMMENT_PREFIX, '' );
```

Intermediate is not necessary.

kgryte committed:
refactor: avoid unnecessary intermediate variable

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
docs: update description

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
docs: update description

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (changes requested):
Only a few comments remain wrt the main implementation.

kgryte commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into philipp/add-typescript-doctest-linting

Planeshifter committed:
chore: address PR feedback

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter committed:
chore: remove backtick and add missing empty line

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter commented on lib/node_modules/@stdlib/_tools/eslint/rules/tsdoc-declarations-doctest/test/fixtures/packages/@stdlib/ndarray/vector/int32/lib/index.js:
```diff
@@ -0,0 +1,29 @@
+
+var Int32Vector = require( '@stdlib/ndarray/vector/int32' );
+
+// EXPORTS //`
```
Fixed. I don't recall enabling Cursor Bugbot so a bit surprised why we got this comment here.

kgryte replied:
That is a bit odd. I didn't enable it either. @headlessNode Do you know anything about this?

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/tsdoc-declarations-doctest/lib/add_package_to_scope.js:
```diff
@@ -0,0 +1,86 @@
+		}
+		interfaceMatch = sourceText.match( RE_DECLARE_VAR_INTERFACE );
+		if ( interfaceMatch ) {
+			// Make the function available under both the variable and interface names:
```
This is a little bit odd, but I understand why this is being done this way. Namely, `ctor` is a dummy variable name in the example used to illustrate the need for this check. What we actually care about is `Int32Vector`.

My sense is this could break at some point, where we double alias, which then clashes with a real alias. But we can cross that bridge when we get there.

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/tsdoc-declarations-doctest/lib/add_package_to_scope.js:
```diff
@@ -0,0 +1,86 @@
+				scope[ namespaceMatch[1] ] = pkg;
+			}
+		}
+		// For all non-function types (including objects), also check const pattern (e.g., `declare const PI: number;` or `declare const obj: {...};`). Note: intentional fall-through to handle both patterns for objects...
```
```suggestion
			}
			// Note: intentional fall-through to handle multiple patterns for objects...
		}
		// For all non-function types (including objects), also check const pattern (e.g., `declare const PI: number;` or `declare const obj: {...};`)...
```

kgryte committed:
docs: update comments

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (approved):
LGTM. Let's get this in, and we can iterate from here.
